src/Pact/Types/Command.hs

```diff
@@ -253,37 +254,48 @@ instance FromJSON UserSig where
   {-# INLINE parseJSON #-}
 
 
+data CommandResult = CommandResult {
+  _crReqKey :: RequestKey,
+  _crTxId :: Maybe TxId,
+  _crResult :: Value
+  } deriving (Eq,Show)
 
+-- | Actual value of a `CommandResult`.
+data CommandValue a
+  = CommandSuccess a
+  | CommandFailure CommandError
+  deriving (Eq, Show)
 
-
-
+-- | Error details for `CommandValue`.
 data CommandError = CommandError {
       _ceMsg :: String
     , _ceDetail :: Maybe String
-}
-instance ToJSON CommandError where
-    toJSON (CommandError m d) =
-        object $ [ "status" .= ("failure" :: String)
-                 , "error" .= m ] ++
-        maybe [] ((:[]) . ("detail" .=)) d
-
-newtype CommandSuccess a = CommandSuccess { _csData :: a }
-  deriving (Eq, Show)
-
-instance (ToJSON a) => ToJSON (CommandSuccess a) where
-    toJSON (CommandSuccess a) =
-        object [ "status" .= ("success" :: String)
-               , "data" .= a ]
-
-instance (FromJSON a) => FromJSON (CommandSuccess a) where
-    parseJSON = withObject "CommandSuccess" $ \o ->
+  } deriving (Eq, Show)
+
+
+instance ToJSON a => ToJSON (CommandValue a) where
+    toJSON = \case
+      CommandSuccess a -> object
+        [ "status" .= ("success" :: Text)
+        , "data"   .= a
+        ]
+      CommandFailure (CommandError {..}) -> object $
+        [ "status" .= ("failure" :: Text)
+        , "error" .= _ceMsg
+        ]
+        ++ maybe [] (pure . ("detail" .=)) _ceDetail
+
+instance FromJSON a => FromJSON (CommandValue a) where
+  parseJSON = withObject "CommandValue" $ \o -> do
+    status <- o .: "status"
+    case status of
+      "failure" -> do
+        msg <- o .: "error"
+        detail <- o .:? "detail"
+        pure $ CommandFailure $ CommandError msg detail
+      "success" -> do
         CommandSuccess <$> o .: "data"
-
-data CommandResult = CommandResult {
-  _crReqKey :: RequestKey,
-  _crTxId :: Maybe TxId,
-  _crResult :: Value
-  } deriving (Eq,Show)
+      _ -> fail $ "Expected a status of `success` or `failure`, not: " <> status
```

Comment:
Do we need this? We control our spec - I'm not sure this case is necessary

Reply:
Well it for sure will not do any harm. I get what you mean though, a `CommandValue` currently does not come from the outside ... although sure it does - for `pact-web`! So if the wire format changes for example this would be a state that could easily be reached. In any case, I like to be on the safe(r) side - having a descriptive error instead of a random crash.